### PR TITLE
Support more cases in rewriting of ConvTranspse op

### PR DESCRIPTION
**Pull request by imaihal**
Subsequent PR of https://github.com/onnx/onnx-mlir/pull/1953 to support auto_pad of SAME_UPPER/LOWER for ConvTranspose op.
Also removed redundant code.

- [x] Use ONNXOpBuiler in more ops and remove result type when using it.
- [x] Update shape inference to use latest style.
- [x] Support auto_pad SAME_UPPER/SAME_LOWER.

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps/NN/Conv.cpp [hunk 1]
```diff
@@ -741,138 +670,14 @@ LogicalResult ONNXConvTransposeOp::inferShapes(
...
-  outputShapeFinal.emplace_back(outChannels);
-  // Insert spatial dims.
-  for (int32_t i = 0; i < spatialRank; ++i) {
-    outputShapeFinal.emplace_back(ArrayAttrIntVal(outputShape, i));
-  }
-  getResult().setType(
-      RankedTensorType::get(outputShapeFinal, xTy.getElementType()));
-
-  return success();
+  Type elementType = getX().getType().cast<ShapedType>().getElementType();
```
Comment (tungld) [inline src/Dialect/ONNX/ONNXOps/NN/Conv.cpp:678]: You can use `getElementType(getX().getType())` from `src/Support/TypeUtilities.hpp`.

File: src/Dialect/ONNX/DialectBuilder.cpp [hunk 2]
```diff
@@ -146,18 +146,19 @@ Value OnnxBuilder::mul(Type resultType, Value A, Value B) const {
       resultType, toTensor(A), toTensor(B));
 }
 
-Value OnnxBuilder::pad(Type outputType, Value input, Value pads,
-    Value constantValue, std::string mode) const {
+Value OnnxBuilder::pad(
+    Value input, Value pads, Value constantValue, std::string mode) const {
+  Type elementType = input.getType().cast<ShapedType>().getElementType();
```
Comment (tungld) [inline src/Dialect/ONNX/DialectBuilder.cpp:151]: You can use `getElementType(input.getType())` from `src/Support/TypeUtilities.hpp`.

File: src/Transform/ONNX/Decompose.cpp [hunk 3]
```diff
@@ -270,40 +271,46 @@ Value reverseWeightTensor(
...
-}
-
-ArrayAttr createUnitStrides(PatternRewriter &rewriter, ArrayAttr strides) {
-  SmallVector<int64_t, 2> unitStrides(strides.size(), 1);
-  return rewriter.getI64ArrayAttr(unitStrides);
+bool hasUnitStrides(ONNXConvTransposeOp op) {
+  ONNXConvTransposeOpShapeHelper shapeHelper(op.getOperation(), {});
+  LogicalResult shapecomputed = shapeHelper.computeShape();
+  assert(succeeded(shapecomputed) &&
+         "unexpected inferShapes failuer for ConvTrans op");
```
Comment (tungld) [inline src/Transform/ONNX/Decompose.cpp:311]: Please use `computeShapeAndAssertOnFailure` instead of `computeShape`, that includes assertion.

Comment (tungld) [reply src/Transform/ONNX/Decompose.cpp:311]: Do we really need to call shapeHelper just for getting the stride attribute?

File: src/Transform/ONNX/Decompose.cpp [hunk 4]
```diff
@@ -397,47 +363,67 @@ Value insertPadAxis(PatternRewriter &rewriter, Location loc, Value input,
...
 Value insertAdditionalPadsConvTranspose(PatternRewriter &rewriter, Location loc,
-    ONNXConvOp op, Value input, ArrayAttr outputShapeAttr) {
-  ONNXConvOpShapeHelper shapeHelper(op.getOperation(), {});
-  shapeHelper.computeShapeAndAssertOnFailure();
-  int inputRank = shapeHelper.getOutputDims().size();
+    ONNXConvOp convOp, Value input, ONNXConvTransposeOp op) {
+  ONNXConvOpShapeHelper convShapeHelper(convOp.getOperation(), {});
+  Type elementType = input.getType().cast<ShapedType>().getElementType();
+  assert(succeeded(convShapeHelper.computeShapeAndUpdateType(elementType)) &&
+         "unexpected inferShapes failure for Conv op");
```
Comment (tungld) [inline src/Transform/ONNX/Decompose.cpp:398]: `computeShapeAndUpdateType` has assertion. There is no need to assert here.

File: src/Transform/ONNX/Decompose.cpp [hunk 5]
```diff
@@ -397,47 +363,67 @@ Value insertPadAxis(PatternRewriter &rewriter, Location loc, Value input,
...
-                    ? shapeHelper.getOutputDims()[i].getLiteral()
+    int64_t d = convShapeHelper.getOutputDims()[i].isLiteral()
+                    ? convShapeHelper.getOutputDims()[i].getLiteral()
                     : ShapedType::kDynamic;
     inputShape.emplace_back(d);
   }
+  ONNXConvTransposeOpShapeHelper shapeHelper(op.getOperation(), {});
+  LogicalResult shapecomputed = shapeHelper.computeShape();
+  assert(succeeded(shapecomputed) &&
+         "unexpected inferShapes failuer for ConvTrans op");
```
Comment (tungld) [inline src/Transform/ONNX/Decompose.cpp:410]: Please use `computeShapeAndAssertOnFailure` instead.

**Review by tungld — CHANGES_REQUESTED** [review]
Thanks for refactoring the code for ConvTransposeOp! LGTM in general.

I see that shapeHelper is called again and again in different functions. I guess it's because of the use of tablegen, so you cannot pass it around. So, to avoid calling shapeHelper multiple times, it's perhaps better to write the pattern in C++ instead of tablegen. What do you think?

File: src/Dialect/ONNX/ONNXOps/NN/Conv.cpp (hunk 1 above)
Comment (imaihal) [reply src/Dialect/ONNX/ONNXOps/NN/Conv.cpp]: Done. I fixed the same pattern in the file. Thanks!

File: src/Dialect/ONNX/DialectBuilder.cpp (hunk 2 above)
Comment (imaihal) [reply src/Dialect/ONNX/DialectBuilder.cpp]: Done. Thanks!

File: src/Transform/ONNX/Decompose.cpp [hunk 6]
```diff
@@ -270,40 +271,46 @@ Value reverseWeightTensor(
...
+ArrayAttr getPadsConvTranspose(
+    PatternRewriter &rewriter, Location loc, ONNXConvTransposeOp op) {
+  // Calculate pads for generated Conv op.
   // new_pads = kernel -  pads - 1
   // Reference: Dumoulin, Vincent, and Francesco Visin. "A guide to convolution
   // arithmetic for deep learning." arXiv preprint arXiv:1603.07285 (2016).
+  ONNXConvTransposeOpShapeHelper shapeHelper(op.getOperation(), {});
+  LogicalResult shapecomputed = shapeHelper.computeShape();
+  assert(succeeded(shapecomputed) &&
+         "unexpected inferShapes failuer for ConvTrans op");
```
[the remark this answers is not in the document]
Comment (imaihal) [reply src/Transform/ONNX/Decompose.cpp:287]: Done. Thanks!

File: src/Transform/ONNX/Decompose.cpp (hunk 3 above)
Comment (imaihal) [reply src/Transform/ONNX/Decompose.cpp]: Removed shapehelper in this function. Thanks.

File: src/Transform/ONNX/Decompose.cpp (hunk 4 above)
Comment (imaihal) [reply src/Transform/ONNX/Decompose.cpp]: Done. Thanks.

File: src/Transform/ONNX/Decompose.cpp (hunk 5 above)
Comment (imaihal) [reply src/Transform/ONNX/Decompose.cpp]: Done. Thanks!

File: src/Transform/ONNX/Decompose.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tungld) [reply src/Transform/ONNX/Decompose.cpp:287]: @imaihal I still see the old code here. Did you push your change to the PR?

[the remark this answers is not in the document]
Comment (imaihal) [reply src/Transform/ONNX/Decompose.cpp]: @tungld Sorry. I may forgot to include the change.

**Comment by imaihal** [comment]
> So, to avoid calling shapeHelper multiple times, it's perhaps better to write the pattern in C++ instead of tablegen. What do you think?

I think so... I used tablegen to use the same approach with other rewriting, but most of code for this rewriting are written in C++ functions.

**Comment by tungld** [comment]
>I think so... I used tablegen to use the same approach with other rewriting, but most of code for this rewriting are written in C++ functions.

Please let us know if you are moving to C++ code or still want to keep tablegen code. Thanks!

**Comment by imaihal** [comment]
@tungld Thanks for letting me know. I want to set it as TODO, but is it acceptable?

**Review by tungld — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**